src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -498,9 +528,93 @@ def _match_account(account, subscription_id, secondary_key_name, secondary_key_v
         s = next((x for x in subscriptions if x.get(_STATE) == 'Enabled'), None)
         return s or subscriptions[0]
 
+    @staticmethod
+    def _interactively_select_subscription(subscriptions, active_one):
+        from azure.cli.core.style import format_styled_text, Style
+        index_to_subscription_map = {}
+        table_data = []
+        subscriptions_sorted = sorted(subscriptions, key=lambda s: s[_SUBSCRIPTION_NAME].lower())
+
+        def get_tenant_string(subscription):
+            try:
+                return subscription[_TENANT_DISPLAY_NAME]
+            except KeyError:
+                return subscription[_TENANT_ID]
+
+        def highlight_text(text, is_default):
+            return format_styled_text((Style.HIGHLIGHT, text)) if is_default else text
+
+        for index, sub in enumerate(subscriptions_sorted, start=1):
+            # There is no need to use int, as int requires parsing. str match is sufficient.
+            index_str = str(index)  # '1', '2', ...
+            index_to_subscription_map[index_str] = sub
+
+            # asterisk = format_styled_text((Style.WARNING, '*'))
+            asterisk = ' *'
+            is_default = sub is active_one
+
+            # Trim subscription name if it is too long
+            subscription_name = sub[_SUBSCRIPTION_NAME]
+            length_limit = 40
+            if len(subscription_name) > length_limit:
+                subscription_name = subscription_name[:length_limit - 3] + '...'
+
+            row = {
+                'No': f'[{index_str}]' + (asterisk if is_default else ''),
+                'Subscription name': highlight_text(subscription_name, is_default),
+                'Subscription ID': highlight_text(sub[_SUBSCRIPTION_ID], is_default),
+                'Tenant': highlight_text(get_tenant_string(sub), is_default)
+            }
+            table_data.append(row)
+
+        from tabulate import tabulate
+        table_str = tabulate(table_data, headers="keys", tablefmt="simple", disable_numparse=True)
+
+        print()
+        print('[Tenant and subscription selection]')
+        print()
+        print(table_str)
+        print()
+        tenant_string = get_tenant_string(active_one)
+        print("The default is marked with an *; "
+              f"the default tenant is '{tenant_string}' and subscription is "
+              f"'{active_one[_SUBSCRIPTION_NAME]}' ({active_one[_SUBSCRIPTION_ID]}).")
+        print()
+
+        from knack.prompting import prompt, NoTTYException
+
+        # Keep prompting until the user inputs a valid index
+        while True:
+            try:
+                select_index = prompt('Select a subscription and tenant (Type a number or Enter for no changes): ')
+            except NoTTYException:
+                # This is a good example showing interactive and non-TTY are not contradictory
+                logger.warning("No TTY to select the default subscription.")
+                return active_one
```

Comment:
If no TTY is available, should we still print the subscription table or the echo message?

Reply:
We probably want to skip the selection process in the case there is no TTY available.

Reply:
I agree, but should we still show the interactive interface, or switch to the previous behavior of returning a JSON?

Reply:
If `stdin` is not a TTY, the user may be running `echo 1 | az login`. This is unlikely as the user can't know beforehand which subscription will be the one they want to select, and the number can change.

If `stdout` is not a TTY, the user may be running `az login | grep`. Not showing the JSON output will be a breaking change in this case.

Therefore, we decide to show interactive subscription selection only when both `stdin` and `stdout` are TTY.